jobrunner/git.py

```diff
@@ -71,6 +71,32 @@ def checkout_commit(repo_url, commit_sha, target_dir):
     )
 
 
+def commit_reachable_from_ref(repo_url, commit_sha, ref):
+    """
+    Given a `ref` (branch name, tag, etc) on a remote repo, check whether the
+    supplied commit is reachable from that ref.
+    """
+    ref_sha = get_sha_from_remote_ref(repo_url, ref)
+    # The easy case and the case I expect to be hit almost every time as the UI
+    # currently only supports running against the branch head
+    if commit_sha == ref_sha:
+        return True
+    # However a well (or badly) timed push could cause the target sha and the
+    # branch head to diverge, so we need to handle that. In order to do so we
+    # need to fetch the history of the branch. We first fetch just the last 10
+    # commits on the assumption that it's probably one of those. If that fails
+    # we fetch the entire branch history.
+    repo_dir = get_local_repo_dir(repo_url)
+    ensure_git_init(repo_dir)
+    fetch_commit(repo_dir, repo_url, ref_sha, depth=10)
```

Comment:
Not at all relevant to this PR but these three lines feel like something we do in various places, would an abstraction to them be useful at some point?

Reply:
I had similar feelings initially but we only actually use them twice and in the other context there's a no-op check (`commit_already_fetched`) which is only applicable when `depth=1` so I think it would end up being less readable to try to make an abstraction out of them.